common/journald/src/stream.rs

```diff
@@ -0,0 +1,356 @@
+            // stream to prevent a race condition
+            drop(sender);
+            call_waker();
+        });
+
+        self.thread = Some(thread);
+        self.receiver = Some(receiver);
+    }
+
+    fn drop_thread(&mut self) {
+        self.shared_state.lock().unwrap().is_alive = false;
+        if let Some(thread) = self.thread.take() {
+            if let Err(e) = thread.join() {
+                warn!("unable to join journald's worker thread: {:?}", e)
+            }
+        }
+    }
+}
+
+impl FutureStream for Stream {
+    type Item = Vec<LineBuilder>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        let mut self_ = self.as_mut();
+
+        if let Some(ref receiver) = self_.receiver {
+            match receiver.try_recv() {
+                Ok(line) => {
+                    return Poll::Ready(Some(vec![line]));
+                }
+                Err(TryRecvError::Disconnected) => {
+                    warn!("journald's main thread unable to read from worker thread, restarting worker thread...");
+                    self_.drop_thread();
+                    self_.spawn_thread();
+                }
+                _ => {}
+            }
+        } else {
+            warn!(
+                "journald's main thread missing connection to worker thread, shutting down stream"
+            );
+            return Poll::Ready(None);
+        }
+
+        let mut shared_state = self_.shared_state.lock().unwrap();
+        shared_state.waker = Some(cx.waker().clone());
+        Poll::Pending
+    }
+}
+
+impl Drop for Stream {
+    fn drop(&mut self) {
+        self.drop_thread();
+    }
+}
+
+struct Reader {
+    reader: Journal,
+}
+
+impl Reader {
+    fn new(path: Path) -> Self {
+        let mut reader = match path {
+            Path::Directory(path) => Journal::open_directory(&path, JournalFiles::All, false)
+                .expect("Could not open journald reader for directory"),
+            Path::Files(paths) => {
+                let paths: Vec<&std::path::Path> = paths.iter().map(PathBuf::as_path).collect();
+                Journal::open_files(&paths).expect("Could not open journald reader for paths")
+            }
+        };
+        reader
+            .seek(JournalSeek::Tail)
+            .expect("Could not seek to tail of journald logs");
+
+        Self { reader }
+    }
+
+    fn process_next_record(&mut self) -> RecordStatus {
+        let record = match self.reader.next_entry() {
+            Ok(Some(record)) => record,
+            Ok(None) => return RecordStatus::NoLines,
+            Err(e) => panic!("Unable to read next record from journald: {}", e),
+        };
+
+        let timestamp = match self.reader.timestamp() {
+            Ok(timestamp) => Local
+                .timestamp(
+                    timestamp.duration_since(UNIX_EPOCH).unwrap().as_secs() as i64,
```

Comment:
I don't think we can always unwrap this, especially if we ever look back to the start of the log. It's possible that the system would come up with it's state some time in the future, the get set to the correct time by ntpd. Between boot and ntpd setting things straight logs could be generated, then when we start reading the logs would look like they were in the future and this would panic.

We can't do anything about the dodgy timestamps, but we could still handle this error and warn! that we have sent some logs from the future